rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,379 @@
+# keras-nlp Transformer Encoder API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com), Hongkun Yu (hongkuny@google.com)|
+| **Sponsor(s)** | Mark Omernick (momernick@google.com)|
+| **Updated**   | 2020-09-21                                           |
+
+
+## Objective
+
+We aim at providing a set of Keras layers to handle Transformer-Encoder BERT-style models.
+
+## Key Benefits
+
+BERT-style Transformer-Encoders are a state-of-art technique that powers many NLP tasks:
+
+- Single sentence classification task, e.g., sentiment analysis
+- Sentence pair classification task, e.g., next sentence prediction
+- Question answering task, e.g., SQuAD
+- Single sentence tagging task, e.g., named entity recognition
+
+With this proposal, Keras users will be able to handle the tasks above with a simple API. 
+
+## Design overview
+
+This proposal builds on the assumption that inputs are lookup indices, i.e., `tf.int64` sequences.
+Tokenization is not part of this proposal but will be our immediate next step.
+
+### Classification task
+
+Case where a user want to use a pretrained BERT encoder for sentiment analysis:
+
+```python
+# Considering a imbd review dataset
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_nlp
+import tensorflow_text as tftext
+
+imdb_reviews = tfds.load('imdb_reviews')
+train_ds = imdb_reviews['train'].batch(32)
+test_ds = imdb_reviews['test'].batch(32)
+
+# Tokenization with BertTokenizer
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
+  token_ids = tokenizer.tokenize(input_text)
+  cls = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([101], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  sep = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([102], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  token_ids = tf.concat([cls, token_ids, sep], axis=1)
+  segment_ids = tf.concat([tf.zeros_like(cls), tf.ones_like(token_ids), tf.ones_like(sep)], axis=1)
+  output_shape = [None, SEQUENCE_LENGTH]
+  token_ids = token_ids.merge_dims(-2, -1)
+  segment_ids = segment_ids.merge_dims(-2, -1).to_tensor(shape=output_shape)
+  input_mask = tf.ones_like(token_ids).to_tensor(shape=output_shape)
+  token_ids = token_ids.to_tensor(shape=output_shape)
+  return {
+      'input_ids': token_ids,
+      'input_mask': input_mask,
+      'input_type_ids': segment_ids
+  }
+
+strategy = tf.distribute.TPUStrategy(...)
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size=30522, max_sequence_length=512, type_vocab_size=2)
+  checkpoint_path = "gs://<bucket_name>/<file_path>"
+  checkpoint = tf.train.Checkpoint(model=encoder)
+  checkpoint.restore(checkpoint_path).assert_consumed().run_restore_ops()
```

Comment:
The 3 lines above (involving two classes and 3 methods) should just be `model.load_weights(checkpoint_path)`.

Reply:
Done